From 8a9121bb11f0423aefe5a51252986765e10214b0 Mon Sep 17 00:00:00 2001
From: Olivier Fourdan <ofourdan@redhat.com>
Date: Fri, 8 Nov 2024 18:18:32 +0100
Subject: [PATCH 5/5] compositor: Make set-above configurable

GNOME Kiosk has heuristics to place windows on the top group.

However, in some cases, it may desirable to avoid that behavior as that
prevents the windows stacking from being changed by the user.

Add a new option "set-above" to the configuration file to control that
behaviour.

If not specified, the existing heuristics will be applied, not breaking
possible existing setups.

Closes: https://gitlab.gnome.org/GNOME/gnome-kiosk/-/issues/26
(cherry picked from commit b50c25fe739c00f4a1984919ee93cd351f58dfc1)
---
 CONFIG.md                     | 3 +++
 compositor/kiosk-compositor.c | 9 +++++++++
 window-config.ini             | 2 ++
 3 files changed, 14 insertions(+)

diff --git a/CONFIG.md b/CONFIG.md
index b0045ee..652fb71 100644
--- a/CONFIG.md
+++ b/CONFIG.md
@@ -42,6 +42,7 @@ The following "*set*" keys are supported:
  * `set-y` (integer) - the Y position
  * `set-width` (integer) - the width
  * `set-height` (integer) - the height
+ * `set-above` (boolean) - Whether the window should be placed on a layer above
 
 # Example
 
@@ -51,6 +52,8 @@ The following "*set*" keys are supported:
   set-x=0
   set-y=0
   set-fullscreen=false
+  # The following will place all windows on the same layer
+  set-above=false
 
   # Make all Mozilla windows fullscreen
   [mozilla]
diff --git a/compositor/kiosk-compositor.c b/compositor/kiosk-compositor.c
index 650ac50..aa5603d 100644
--- a/compositor/kiosk-compositor.c
+++ b/compositor/kiosk-compositor.c
@@ -385,6 +385,15 @@ static gboolean
 kiosk_compositor_wants_window_above (KioskCompositor *self,
                                      MetaWindow      *window)
 {
+        gboolean set_above;
+
+        if (kiosk_window_config_get_boolean_for_window (self->kiosk_window_config,
+                                                        window,
+                                                        "set-above",
+                                                        &set_above))
+                return set_above;
+
+        /* If not specified in the config, use the heuristics */
         if (meta_window_is_screen_sized (window)) {
                 return FALSE;
         }
diff --git a/window-config.ini b/window-config.ini
index c4c825e..f4f37b3 100644
--- a/window-config.ini
+++ b/window-config.ini
@@ -7,6 +7,8 @@
 [all]
 set-x=0
 set-y=0
+# Uncomment the following to have all windows on the same layer
+# set-above=false
 
 # Make all Mozilla windows fullscreen
 [browser]
-- 
2.49.0

